### ISSUE TITLE
[SUBMIT] Replacing strapi default WYSIWYG editor (draftjs) with Tinymce editor

### ISSUE THREAD

**Issue by AmerAlahmar** [strapi/community-content#180]
**Replacing strapi default WYSIWYG editor (draftjs) with Tinymce editor**

I know there's an [article](https://strapi.io/blog/how-to-change-the-wysiwyg-in-strapi) on how to replace Strapi's editor with CKEditor but it seems outdated, I also know about the one in the [docs](https://strapi.io/documentation/v3.x/guides/registering-a-field-in-admin.html#introduction)  which should be a good start but it's specific to CKEditor.

I have seen a few unanswered questions on how to replace it with Tinymce Editor in Strapi blog and in Tinymce's  while I was trying to implement it myself.

So I think it might be useful for beginners like me (specially for React newbies).

**Format**

- [x] Article

**What do you guy's think?**
❤️ I love it!
🚀 I can help you!

**Comment by AmerAlahmar** [comment]
**I have prepared a quick draft that includes the steps and the commands needed, I just need to:**
- put it in nice shape
- include more explanations of the steps.

Here is the draft _(matches Strapi docs style)_ :

1. Create a new project:
Create an application using SQLite and prevent the server from starting automatically as we will create a plugin right after the project generation

```
#YARN:
yarn create strapi-app my-app --quickstart --no-run

#NPX:
npx create-strapi-app my-app --quickstart --no-run

#STRAPI:
strapi new my-app --quickstart --no-run
```

2. Generate a plugin:
Generate a new plugin  ```wysiwyg```

```
cd my-app

#YARN:
yarn strapi generate:plugin wysiwyg

#NPM:
npm run strapi generate:plugin wysiwyg

#STRAPI:
strapi generate:plugin wysiwyg
```

3. Install the needed dependencies:
install ```tinymce-react``` package

```
cd plugins/wysiwyg

#YARN:
yarn add @tinymce/tinymce-react

#NPM:
npm install @tinymce/tinymce-react
```

4. Start your application with the front-end development mode:
Go back to to strapi root folder and start in development mode and add the option ```--watch-admin```

```
cd ../..

#YARN:
yarn develop --watch-admin

#NPM:
npm run develop --watch-admin

#STRAPI:
strapi develop --watch-admin
```

5. Create the MediaLib:
Path — ./plugins/wysiwyg/admin/src/components/MediaLib/index.js

```
import React, { useEffect, useState } from 'react';
import { useStrapi, prefixFileUrlWithBackendUrl } from 'strapi-helper-plugin';
import PropTypes from 'prop-types';

const MediaLib = ({ isOpen, onChange, onToggle }) => {
  const {
    strapi: {
      componentApi: { getComponent },
    },
  } = useStrapi();
  const [data, setData] = useState(null);
  const [isDisplayed, setIsDisplayed] = useState(false);

  useEffect(() => {
    if (isOpen) {
      setIsDisplayed(true);
    }
  }, [isOpen]);

  const Component = getComponent('media-library').Component;

  const handleInputChange = data => {
    if (data) {
      const { url } = data;

      setData({ ...data, url: prefixFileUrlWithBackendUrl(url) });
    }
  };

  const handleClosed = () => {
    if (data) {
      onChange(data);
    }

    setData(null);
    setIsDisplayed(false);
  };

  if (Component && isDisplayed) {
    return (
      <Component
        allowedTypes={['images', 'videos', 'files']}
        isOpen={isOpen}
        multiple={false}
        noNavigation
        onClosed={handleClosed}
        onInputMediaChange={handleInputChange}
        onToggle={onToggle}
      />
    );
  }

  return null;
};

MediaLib.defaultProps = {
  isOpen: false,
  onChange: () => {},
  onToggle: () => {},
};

MediaLib.propTypes = {
  isOpen: PropTypes.bool,
  onChange: PropTypes.func,
  onToggle: PropTypes.func,
};

export default MediaLib;
```

6. Create the WYSIWYG Wrapper:
Path — ./plugins/wysiwyg/admin/src/components/Wysiwyg/index.js

```
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';
import { Button } from '@buffetjs/core';
import { Label, InputDescription, InputErrors } from 'strapi-helper-plugin';
import TinyEditor from '../Tinymce';
import MediaLib from '../MediaLib';

const Wysiwyg = ({
  inputDescription,
  errors,
  label,
  name,
  noErrorsDescription,
  onChange,
  value,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  let spacer = !isEmpty(inputDescription) ? <div style={{ height: '.4rem' }} /> : <div />;

  if (!noErrorsDescription && !isEmpty(errors)) {
    spacer = <div />;
  }

  const handleChange = data => {
    if (data.mime.includes('image')) {
      const imgTag = `<p><img src="${data.url}" caption="${data.caption}" alt="${data.alternativeText}"></img></p>`;
      const newValue = value ? `${value}${imgTag}` : imgTag;

      onChange({ target: { name, value: newValue } });
    }

    // Handle videos and other type of files by adding some code
  };

  const handleToggle = () => setIsOpen(prev => !prev);

  return (
    <div
      style={{
        marginBottom: '1.6rem',
        fontSize: '1.3rem',
        fontFamily: 'Lato',
      }}
    >
      <Label htmlFor={name} message={label} style={{ marginBottom: 10 }} />
      <div>
        <Button color="primary" onClick={handleToggle}>
          MediaLib
        </Button>
      </div>
///////////////////////////Tinymce Editor//////////////////////////////
      <TinyEditor name={name} onChange={onChange} value={value} />
//////////////////////////////////////////////////////////////////////
      <InputDescription
        message={inputDescription}
        style={!isEmpty(inputDescription) ? { marginTop: '1.4rem' } : {}}
      />
      <InputErrors errors={(!noErrorsDescription && errors) || []} name={name} />
      {spacer}
      <MediaLib onToggle={handleToggle} isOpen={isOpen} onChange={handleChange} />
    </div>
  );
};

Wysiwyg.defaultProps = {
  errors: [],
  inputDescription: null,
  label: '',
  noErrorsDescription: false,
  value: '',
};

Wysiwyg.propTypes = {
  errors: PropTypes.array,
  inputDescription: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.shape({
      id: PropTypes.string,
      params: PropTypes.object,
    }),
  ]),
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.shape({
      id: PropTypes.string,
      params: PropTypes.object,
    }),
  ]),
  name: PropTypes.string.isRequired,
  noErrorsDescription: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
};

export default Wysiwyg;
```

7. Implemente Tinymce:
Path — ./plugins/wysiwyg/admin/src/components/Tinymce/index.js

```
import React from "react";
import PropTypes from "prop-types";
import { Editor } from "@tinymce/tinymce-react";

const TinyEditor = ({ onChange, name, value }) => {

  return (
    <Editor
      apiKey="YOUR API KEY"
      value={value}
      tagName={name}
      onEditorChange={(editorContent) => {
        onChange({ target: { name, value: editorContent } });
      }}
      outputFormat='text'
      init={{
        /*NOTE: It's better NOT to allow the user to insert images from device or external sources as IT WILL MAKE A MESS*/
        plugins: "YOUR PLUGINS",
        //other configurations...
      }}
    />
  );
};


TinyEditor.propTypes = {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
};

export default TinyEditor;
```

8. rebuild strapi:

```
#YARN:
yarn build

#NPM:
npm run build

#STRAPI:
strapi build
```

**Comment by malgamves** [comment]
This looks really good @AmerAlahmar - sorry for catching it late. I like where this is heading! Hows the progress on explanations and context?

**Comment by malgamves** [comment]
Hey @AmerAlahmar how is it going with this?

**Comment by Mcastres** [comment]
I an write explanations and context if you want @AmerAlahmar

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar Been a while. Thank you for putting your time on this. Are you still interested in finishing this article?

**Comment by AmerAlahmar** [comment]
Hey @malgamves @ShadaW11 
I am deeply sorry for overlooking this all this time, here is the current situation:

After a few tries and assessments of Strapi,  unfortunately, we decided not to go with Strapi for now (while keeping an eye on it). For that reason didn't get deeper with this.

HOWEVER, I still intend to finish what I started, but it will take some time I am afraid because I will have to re-install Strapi and see if the steps are still working with the newest version and then of course write the article ✍🏽 

@Mcastres Any help would be very much welcomed 😃

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar Thank you for getting back to me so quickly. No worries at all. Thanks again for explaining the situation. I totally understand and happy to know that you intend on finishing the article. We are also ready to help you if you have any questions. Also, if you have time, could you please check and register on the ["write for community"](https://strapi.io/write-for-the-community) page. We have updated our process (writing guidelines, payment,..)

**Comment by AmerAlahmar** [comment]
Hi @ShadaW11 @malgamves,
How are you doing? hope everything is fine.

I am offended to see that you have added the "looking for author" label!!
Just kidding :smiley: I have been stalling for quite some time now, I deserve it :grin: .

Anyway, I have just finished writing the [Tutorial](https://github.com/AmerAlahmar/AmerAlahmar.github.io/blob/main/strapi/replace-default-wysiwyg-editor-with-tinymce.md).

I have spotted quite a few typos and I still need to proofread it, hopefully on the weekend.

However, I am not sure where to go from there, right now the file is in my personal repo so it would be appreciated if you can point me to the submitting instructions.

Thanks in advance, have a nice day.

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar Awesomeee, apologies I thought you were busy but happy to see its done. 

Please register your author details at the bottom of this page [Write for the Community](https://strapi.io/write-for-the-community) and could you please check our FAQ to submit the tutorial, review article guidelines, and also please take a look on how to get and at our Invoice submission guidelines.

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar is the tutorial ready for us to review?

**Comment by AmerAlahmar** [comment]
Hi @ShadaW11 
Sorry not yet 😔

**Comment by AmerAlahmar** [comment]
Hey @ShadaW11 

Please take a look at the file [here on Dropbox](https://www.dropbox.com/scl/fi/u082k8iuabc0732lma6zc/How-to-replace-Strapi-s-default-Wysiwyg-editor-with-Tinymce-editor.paper?dl=0&rlkey=abo9ot9vz2r29z8gy9y6bat5t).

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar thank you for submitting the article. Maxime has left you few comments. please check them once possible.

**Comment by AmerAlahmar** [comment]
Hey @ShadaW11 

I checked Maxime's comments and edit it the article. Thanks for your efforts.

**Comment by ShadaW11** [comment]
Great, thanks @AmerAlahmar, @Mcastres  will check your feedback as soon as he can and will get back to you.

**Comment by ShadaW11** [comment]
Hey @AmerAlahmar, thank you for being patient, @Mcastres has checked your edits and let you feedback. Kindly get back to me once ready. Looking forward to publishing your article.

**Comment by AmerAlahmar** [comment]
Hi @ShadaW11 @Mcastres 

I have just finished adding the requested edits. 💪

Thanks.

**Comment by Mcastres** [comment]
LGTM!

**Comment by ShadaW11** [comment]
Hello @AmerAlahmar [your blog post has been published](https://strapi.io/blog/replace-strapi-default-wysiwyg-editor-with-tiny-mce) 🎉🎉 Please take a look at our FAQ to submit your invoice and get paid.

**Comment by AmerAlahmar** [comment]
Hey @ShadaW11
Thanks for being patient with me. So happy to see it published 😍

One thing though, I spotted a few typos 😅 Anyway to fix them? How?

**Comment by ShadaW11** [comment]
@AmerAlahmar thank you for writing the article. Don't forget to follow the guidelines when issuing and submitting the invoice. Regarding the typos, please share those with me here and ill edit the article right away.

**Comment by AmerAlahmar** [comment]
Hey @ShadaW11 
I submitted the invoice today, thanks 😃 

Regarding the typos:

- Under (MediaLib/index.js) section:
  -- Wrong: ... we don't want to let the user (The person who is using the Editor insert media ...
  -- Should Be: ... we don't want to let the user (The person who is using the Editor) insert media...
  -- Reason: Missing close parenthesis
  
- Under (6. Run Strapi:) section:
  -- Wrong: When you run the last command, it will open a new tab in the browser (if it didn't, ...
  -- Should Be: ... (if it didn't, head to localhost:8000/admin) ...
  -- Reason: Missing close parenthesis
  
- Just Before (7. Configure TinyMCE Editor:) section:
  -- Wrong: ... we started Strapi with --``watch-admin ...
  -- Should Be: ... we started Strapi with --watch-admin ...
  -- Reason: Two extra reversed prime symbols (`)
  
- Again, Just Before (7. Configure TinyMCE Editor:) section:
  -- Wrong: ... we will still be able to see the changes we will make do as we are doing them ...
  -- Should Be: ... we will still be able to see the changes we will make as we are making them ...
  -- Reason: Typo (make do)
 
- Every "Tinymce" Should be replaced with "TinyMCE" (Or vice versa). EXCEPT for the ones inside code blocks 
  Reason: To maintain consistency.
